jdbc-dbcp/src/main/java/io/micronaut/configuration/jdbc/dbcp/DatasourceFactory.java

```diff
@@ -56,9 +65,23 @@ public DataSourcePoolMetadata<BasicDataSource> dbcpDataSourcePoolMetadata(
         DbcpDataSourcePoolMetadata dbcpDataSourcePoolMetadata = null;
         DataSource resolved = dataSourceResolver.resolve(dataSource);
 
-        if (resolved instanceof BasicDataSource) {
-            dbcpDataSourcePoolMetadata = new DbcpDataSourcePoolMetadata((BasicDataSource) resolved);
+        if (resolved instanceof BasicDataSource basicDataSource) {
+            dbcpDataSourcePoolMetadata = new DbcpDataSourcePoolMetadata(basicDataSource);
         }
         return dbcpDataSourcePoolMetadata;
     }
+
+    @Override
+    public void onApplicationEvent(DataSourcePasswordChangedEvent event) {
```

Comment:
I think instead of a new event you should extend `RefreshEventListener` and override `getObservedConfigurationPrefixes()` to get the data which will make it work in other contexts like the refresh endpoint and periodic monitoring.

For UCP you should call:

```
public void reconfigureDataSource(Properties configuration) throws SQLException {
```

For Hikari you should call:

```
hikariDataSource.getHikariConfigMXBean().setPassword("newPassword");
hikariDataSource.getHikariPoolMXBean().softEvictConnections();
```

For DBCP:

```
// set the properties
// and then call
pool.evict();
```

Reply:
I wasn't sure if I am able to populate all `Properties` for UCP or can only set password to properties?
Btw, I think this is not needed, since existing connections will have old password and it should work until connections are evicted by the pool and then regenerated with the new password.

As for the event, I thought this will be more convenient because users might have their own logic/events for password change and this might be simpler and more straightforward.

Reply:
it isn't about being easier it is because this change won't work from the refresh management endpoint or wither periodic config refresh mechanisms

Reply:
Ok, I guess will wait this PR https://github.com/micronaut-projects/micronaut-core/pull/11798 to be merged and then for which drivers needed listen to refresh event and I suppose `datasources.*.password` and extract datasource name and update password if needed.